Pin development dependencies & integrate dependabot

jcugat opened the pull request:
Triggered by https://github.com/encode/httpx/pull/1053#issuecomment-653910614 and https://github.com/encode/httpcore/pull/151#issuecomment-673368004. I set it up a while ago in [my fork of httpx](https://github.com/jcugat/httpx/pulls) and seems to work quite well. We can try it here first, and expand it to httpx too if everybody agrees, or remove it if we find any issue.

After this PR is merged dependabot will be enabled to upgrade dependencies once per week: https://docs.github.com/en/github/administering-a-repository/enabling-and-disabling-version-updates

florimondmanca commented on requirements.txt:
```diff
@@ -1,30 +1,31 @@
+pytest-asyncio==0.14.0
+pytest-cov==2.10.1
+pytest-trio==0.6.0
+sortedcontainers==2.1.0
```
I think this is new? Should we add a comment to clarify why it's there?

Also, actually, we can add `--use-feature="2020-resolver"` at the top of this requirements file, and pip will use the resolver when installing requirements…

Will issue a PR for this, since I think it's the way to go rather than pinning sortedcontainers, now that pip _has_ a resolver! :-)

florimondmanca reviewed (dismissed):
Yes when https://github.com/encode/httpcore/pull/162/files#r470961811 is resolved via #163 :-)

tomchristie commented:
Great. What we *might* want here is a `requirements.base` that holds our actual requirements, and a `requirements.txt` that holds the frozen set of *all* requirements + their dependencies? (That's what I'd normally do with a deployed service, so really you need to make sure you're pinning the indirect package dependencies too)

jcugat commented:
@tomchristie I've implemented your suggestion and also added `./scripts/freeze` to keep both files in sync, let me know what you think.

@florimondmanca I've also integrated your changes from #163

j178 commented on scripts/freeze:
```diff
@@ -0,0 +1,21 @@
+"${PIP}" install -r requirements.base
+
+# Argument "--all" is needed to pin "wheel". Exclude our own editable dependency
+"${PIP}" freeze --all | grep -v '^-e' > requirements.txt
+# Add it back
+echo "-e .[http2]" >> requirements.txt
```
Why not let `pip freeze` handle editable dependency?

jcugat replied:
```console
$ pip install -e .[http2]
Obtaining file:///Users/josepcugat/workspace/httpcore
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Collecting sniffio==1.*
  Using cached sniffio-1.1.0-py3-none-any.whl (4.5 kB)
Collecting h11<0.10,>=0.8
  Using cached h11-0.9.0-py2.py3-none-any.whl (53 kB)
Collecting h2==3.*; extra == "http2"
  Using cached h2-3.2.0-py2.py3-none-any.whl (65 kB)
Collecting hpack<4,>=3.0
  Using cached hpack-3.0.0-py2.py3-none-any.whl (38 kB)
Collecting hyperframe<6,>=5.2.0
  Using cached hyperframe-5.2.0-py2.py3-none-any.whl (12 kB)
Installing collected packages: sniffio, h11, hpack, hyperframe, h2, httpcore
  Running setup.py develop for httpcore
Successfully installed h11-0.9.0 h2-3.2.0 hpack-3.0.0 httpcore hyperframe-5.2.0 sniffio-1.1.0
$ pip freeze
h11==0.9.0
h2==3.2.0
hpack==3.0.0
-e git+git@github.com:encode/httpcore.git@26a37dc1d0f6961bb71ce18331f3c3b45c6f928b#egg=httpcore
hyperframe==5.2.0
sniffio==1.1.0
```

It would be pinning `httpcore` to a specific commit of the git repo instead of the current dir.

j178 replied:
How about `pip freeze --exclude-editable` ?

jcugat replied:
Even better! Didn't know about this argument, updated 👍

j178 commented on scripts/install:
```diff
@@ -15,5 +15,6 @@ else
     PIP="pip"
 fi
 
+"$PIP" install -U "pip >= 20.2" setuptools wheel
```
Install `wheel` before others to avoid warnings like:
```
Using legacy 'setup.py install' for autoflake, since package 'wheel' is not installed.
```

j178 commented:
Hi, based on your excellent work, I added two commits:
- Install packge tools, like `setuptools`, `wheel` before others, to avoid warnings like `Using legacy 'setup.py install' for autoflake, since package 'wheel' is not installed.`
- Remove a redundant `pip install -e .` in `scripts/install`

If there is any problem, please feel free to revert them.

jcugat commented:
@j178 yup, those changes look good.

@tomchristie @florimondmanca are you happy with the overall PR now? Should we try it? 🙂

tomchristie commented on scripts/freeze:
```diff
@@ -0,0 +1,22 @@
+"${PIP}" install -U "pip >= 20.2" setuptools wheel
+"${PIP}" install -r requirements.in
+
+# Argument "--all" is needed to pin "wheel"
```
Okay, tho' we're never installing from the pinned version of `wheel`, since we always preemptively use:

`pip install -U pip setuptools wheel`

So I don't think we really need/want `--all` here.

tomchristie commented on requirements.in:
```diff
@@ -0,0 +1,32 @@
+--use-feature="2020-resolver"
```
Grand yup, we can drop this once pip 20.3 is released, due October 2020.

https://pip.pypa.io/en/stable/news/#id26

j178 commented on requirements.txt:
```diff
@@ -25,9 +25,11 @@ hpack==3.0.0
 httptools==0.1.1
 hyperframe==5.2.0
 idna==2.10
+importlib-metadata==1.7.0
```
Run `scripts/freeze` in my dev environment, resutling in four more dependencies, `importlib-metadata`, `jeepney`, `SecretStorage` and `zipp`, I don't know why…

florimondmanca commented:
@jcugat I pushed an update of dependencies to resolve conflicts with master, but also to see for myself what working with this looks like…

Re @j178's comment in https://github.com/encode/httpcore/pull/162#discussion_r472129736 - it looks like we're going to have some issues with multiplatform support, since we can see several transitive deps are installed on Linux (which I assume is your OS) vs macOS (mine) or even Windows, so we can't have one `requirements.txt` that lists everything for everyone (since `pip freeze` just dumps what's installed instead of adding environment markers as well).

florimondmanca commented:
Closing in favor of #184

Thanks a ton for promoting this change @jcugat!

jcugat commented:
So glad to see this solved in #184, didn't realize the issues with different platforms.

Also, apologies to @florimondmanca and @tomchristie for leaving this open for so long. I've been on holidays and did a necessary break from tech, but I'm back with my batteries fully recharged! 🚀
